host-controller/src/main/java/org/jboss/as/host/controller/mgmt/HostControllerRegistrationHandler.java

```diff
@@ -332,9 +354,15 @@ public void execute(final OperationContext context, final ModelNode operation) t
             }
             // Remotely resolve the subsystem versions and create the transformation
             registrationContext.processSubsystems(transformers, extensions);
-            // Now run the read-domain model operation
-            final ReadMasterDomainModelHandler handler = new ReadMasterDomainModelHandler(hostInfo, transformers, domainController.getExtensionRegistry());
-            context.addStep(READ_DOMAIN_MODEL.getOperation(), handler, OperationContext.Stage.MODEL);
+
+            final Transformers.ResourceIgnoredTransformationRegistry ignoredTransformationRegistry;
+            final Resource resource = context.readResourceFromRoot(PathAddress.EMPTY_ADDRESS);
+            final ReadMasterDomainModelUtil.RequiredConfigurationHolder rc = hostInfo.populateRequiredConfigurationHolder(resource, domainController.getExtensionRegistry());
+            ignoredTransformationRegistry = ReadMasterDomainModelUtil.createHostIgnoredRegistry(hostInfo, rc);
+
+            final ReadMasterDomainModelUtil readUtil = ReadMasterDomainModelUtil.readMasterDomainResourcesForInitialConnect(transformers,
+                    transformationInputs, ignoredTransformationRegistry, transformationInputs.getRootResource());
+            context.getResult().set(readUtil.getDescribedResources());
```

Comment:
Anything in the host-controller module can be changed at will. That module is internal to the kernel and no one else has any business depending on it.

Even within modules like that though it's always good though to reduce visibility of things if they are broader than needed. Lots of stuff has to be public though for use in other packages.

Of the "kernel" (i.e. non-Extension, and non-test-related) modules in core, only these expose API usable outside the kernel:

cli
controller
controller-client
embedded
launcher
network
self-contained
server

The controller and server modules need to be broken up to separate out externally usable stuff from internals. Big task. The module.xml for some others, like cli and controller-client, could do with some tweaking to limit what they export.